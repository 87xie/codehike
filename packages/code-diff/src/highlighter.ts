import { IRawTheme } from "vscode-textmate"
import {
  getBasicThemedCode,
  tokenizePlaceholder,
} from "./placeholder"
import {
  Highlighter,
  setCDN,
  getHighlighter,
  IShikiTheme,
  IThemedToken,
  FontStyle,
  Lang,
} from "shiki"
import { useAsyncMemo } from "./use-async-memo"
import {
  Tween,
  FullTween,
  mapWithDefault,
} from "@code-hike/utils"

type CodeToken = [
  string,
  {
    className?: string
    style?: { color: string | undefined }
  }
]
type CodeLine = CodeToken[]

let highlighterPromise: Promise<Highlighter> | null = null
let highlighter: Highlighter | null = null

export function useHighlighter(
  code: Tween<string>,
  lang: string,
  theme: IRawTheme
): {
  lines: FullTween<CodeLine[]>
  bg: string
  fg: string
} {
  return useAsyncMemo(
    {
      loader: () => highlight(code, lang, theme),
      placeholder: () => {
        const { bg, fg } = getBasicThemedCode("", theme)
        const lines = mapWithDefault(
          code,
          "",
          tokenizePlaceholder
        )
        return { lines, bg, fg }
      },
    },
    [code.prev, code.next, lang, theme]
  )
}

async function highlight(
  code: Tween<string>,
  lang: string,
  theme: IRawTheme
) {
  if (highlighterPromise === null) {
    setCDN("https://unpkg.com/shiki/")
    highlighterPromise = getHighlighter({
      theme: theme as IShikiTheme,
      langs: [lang as Lang], // TODO change lang from string to Lang
    })
  }

  if (highlighter === null) {
    highlighter = await highlighterPromise
  }
  if (missingTheme(highlighter, theme)) {
    await highlighter.loadTheme(theme as IShikiTheme)
  }
  if (missingLang(highlighter, lang)) {
    await highlighter.loadLanguage(lang as Lang)
  }

  const { fg, bg } = highlighter.getTheme(theme.name)

  const lines = mapWithDefault(code, "", code =>
    getCodeLines(highlighter!, code, lang, theme)
  )

  return { lines, bg, fg }
}

function getCodeLines(
  highlighter: Highlighter,
  code: string,
  lang: string,
  theme: IRawTheme
): CodeLine[] {
  const lines = highlighter.codeToThemedTokens(
    code,
    lang,
    theme.name,
    {
      includeExplanation: false,
    }
  )

  return lines.map(line =>
    line.map(token => [
      token.content,
      { style: getStyle(token) },
    ])
  )
}

function missingTheme(
  highlighter: Highlighter,
  theme: IRawTheme
) {
  return !highlighter
    .getLoadedThemes()
    .some(t => t === theme.name)
}

function missingLang(
  highlighter: Highlighter,
  lang: string
) {
  return !highlighter
    .getLoadedLanguages()
    .some(l => l === lang)
}

const FONT_STYLE_TO_CSS = {
  [FontStyle.NotSet]: {},
  [FontStyle.None]: {},
  [FontStyle.Italic]: { fontStyle: "italic" },
  [FontStyle.Bold]: { fontWeight: "bold" },
  [FontStyle.Underline]: { textDecoration: "underline" },
}

function getStyle(token: IThemedToken) {
  const fontStyle = token.fontStyle
    ? FONT_STYLE_TO_CSS[token.fontStyle]
    : {}
  return {
    color: token.color,
    ...fontStyle,
  }
}
