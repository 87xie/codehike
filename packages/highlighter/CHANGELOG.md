# v0.3.0 (Thu Mar 03 2022)

#### üöÄ Enhancement

- Better sections [#139](https://github.com/code-hike/codehike/pull/139) ([@pomber](https://github.com/pomber))
- Allow empty lang [#103](https://github.com/code-hike/codehike/pull/103) ([@pomber](https://github.com/pomber))
- Use shiki [#77](https://github.com/code-hike/codehike/pull/77) ([@pomber](https://github.com/pomber))

#### üêõ Bug Fix

- merge ([@pomber](https://github.com/pomber))
- Update shiki ([@pomber](https://github.com/pomber))
- Handle invalid language [#126](https://github.com/code-hike/codehike/pull/126) ([@pomber](https://github.com/pomber))
- Handle invalid language ([@pomber](https://github.com/pomber))
- Allow empty lang ([@pomber](https://github.com/pomber))
- Fix multi-annotation focus string [#94](https://github.com/code-hike/codehike/pull/94) ([@pomber](https://github.com/pomber))
- Remove dead code ([@pomber](https://github.com/pomber))
- Merge branch 'next' into shiki ([@pomber](https://github.com/pomber))
- Change react dep ([@pomber](https://github.com/pomber))
- Make highlighter load all languages ([@pomber](https://github.com/pomber))
- Upgrade shiki ([@pomber](https://github.com/pomber))
- Add highlighter package ([@pomber](https://github.com/pomber))

#### Authors: 2

- Nirmalya Ghosh ([@ghoshnirmalya](https://github.com/ghoshnirmalya))
- Rodrigo Pombo ([@pomber](https://github.com/pomber))
