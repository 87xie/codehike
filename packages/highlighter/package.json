{
  "name": "@code-hike/highlighter",
  "version": "0.3.0",
  "main": "dist/index.cjs.js",
  "typings": "dist/index.d.ts",
  "module": "dist/index.esm.js",
  "sideEffects": false,
  "files": [
    "dist"
  ],
  "scripts": {
    "x": "x"
  },
  "devDependencies": {
    "@code-hike/script": "^0.3.0",
    "@types/react": "^17.0.2"
  },
  "dependencies": {
    "@code-hike/utils": "^0.3.0",
    "shiki": "^0.10.1"
  },
  "homepage": "https://codehike.org",
  "repository": "code-hike/codehike",
  "publishConfig": {
    "access": "public"
  },
  "author": "Rodrigo Pombo",
  "license": "MIT",
  "funding": {
    "type": "opencollective",
    "url": "https://opencollective.com/codehike"
  }
}
