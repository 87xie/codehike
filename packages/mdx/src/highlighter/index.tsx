import { Code } from "../utils"
import { highlight as light } from "@code-hike/lighter"

const newlineRe = /\r\n|\r|\n/

export async function highlight({
  code,
  lang,
  theme,
}: {
  code: string
  lang: string
  theme: any // TODO type this
}): Promise<Code> {
  if (lang === "text") {
    const lines = code ? code.split(newlineRe) : [""]
    return {
      lines: lines.map(line => ({
        tokens: [{ content: line, props: {} }],
      })),
      lang,
    }
  }

  const r = await light(code, lang, theme)

  const lines = r.lines.map(line => ({
    tokens: line.map(token => ({
      content: token.content,
      props: { style: token.style },
    })),
  }))

  return { lines, lang }
}
