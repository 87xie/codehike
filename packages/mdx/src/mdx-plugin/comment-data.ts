import { Code } from "../utils"
import { annotationsMap } from "../mdx-client/annotations"

const validKeys = ["focus", ...Object.keys(annotationsMap)]

export function getCommentData(
  line: Code["lines"][0],
  lang: string
) {
  const result = bashLikeLangs.includes(lang)
    ? bashLikeComment(line)
    : otherComment(line)

  if (!result) {
    return {}
  }

  const [, key, focusString, data] = result

  if (!validKeys.includes(key)) {
    return {}
  }

  return {
    key,
    focusString,
    data,
  }
}

const commentRegex = /\/\/\s+(\w+)(\S*)\s*(.*)/
function otherComment(line: Code["lines"][0]) {
  const comment = line.tokens.find(t =>
    t.content.trim().startsWith("//")
  )?.content

  if (!comment) {
    return []
  }

  const result = commentRegex.exec(comment)

  if (!result) {
    return []
  }

  return result
}

const bashLikeLangs = [
  "bash",
  "sh",
  "shell",
  "python",
  "py",
]
const bashLikeCommentRegex = /#\s+(\w+)(\S*)\s*(.*)/
function bashLikeComment(line: Code["lines"][0]) {
  const comment = line.tokens.find(t =>
    t.content.trim().startsWith("#")
  )?.content

  if (!comment) {
    return []
  }

  const result = bashLikeCommentRegex.exec(comment)

  if (!result) {
    return []
  }

  return result
}
