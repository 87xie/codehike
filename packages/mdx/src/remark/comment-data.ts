import { Code } from "../utils"
import { annotationsMap } from "../mdx-client/annotations"

const validKeys = ["focus", ...Object.keys(annotationsMap)]

export function getCommentData(
  line: Code["lines"][0],
  lang: string
) {
  const result = bashLikeLangs.includes(lang)
    ? bashLikeComment(line)
    : otherComment(line)

  if (!result) {
    return {}
  }

  const [, key, focusString, data] = result

  if (!validKeys.includes(key)) {
    return {}
  }

  return {
    key,
    focusString,
    data,
  }
}

function getTextAfter(
  line: Code["lines"][0],
  prefix: string
) {
  const firstIndex = line.tokens.findIndex(t =>
    t.content.trim().startsWith(prefix)
  )
  if (firstIndex === -1) {
    return undefined
  }
  return line.tokens
    .slice(firstIndex)
    .map(t => t.content)
    .join("")
}

const commentRegex = /\/\/\s+(\w+)(\S*)\s*(.*)/
function otherComment(line: Code["lines"][0]) {
  const comment = getTextAfter(line, "//")

  if (!comment) {
    return []
  }

  const result = commentRegex.exec(comment)

  if (!result) {
    return []
  }

  return result
}

const bashLikeLangs = [
  "bash",
  "sh",
  "shell",
  "python",
  "py",
]
const bashLikeCommentRegex = /#\s+(\w+)(\S*)\s*(.*)/
function bashLikeComment(line: Code["lines"][0]) {
  const comment = getTextAfter(line, "#")

  if (!comment) {
    return []
  }

  const result = bashLikeCommentRegex.exec(comment)

  if (!result) {
    return []
  }

  return result
}
