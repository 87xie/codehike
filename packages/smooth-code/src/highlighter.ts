import {
  Highlighter,
  setCDN,
  getHighlighter,
  IShikiTheme,
  IThemedToken,
  FontStyle,
  Lang,
} from "shiki"
import { useAsyncMemo } from "./use-async-memo"
import {
  Tween,
  FullTween,
  mapWithDefault,
} from "@code-hike/utils"
import {
  getThemeDefaultColors,
  EditorTheme,
} from "./themes"

type CodeToken = [
  string,
  {
    className?: string
    style?: { color: string | undefined }
  }
]
type CodeLine = CodeToken[]

let highlighterPromise: Promise<Highlighter> | null = null
let highlighter: Highlighter | null = null

export function useHighlighter(
  code: Tween<string>,
  lang: string,
  theme: EditorTheme
): {
  lines: FullTween<CodeLine[]>
} {
  return useAsyncMemo(
    {
      isReady: isHighlighterReady(lang, theme),
      load: () => loadHighlighter(lang, theme),
      run: () => highlight(code, lang, theme),
      placeholder: () => {
        const lines = mapWithDefault(
          code,
          "",
          tokenizePlaceholder
        )
        return { lines }
      },
    },
    [code.prev, code.next, lang, theme]
  )
}

export function isHighlighterReady(
  lang: string,
  theme: EditorTheme
) {
  return (
    highlighter != null &&
    !missingTheme(highlighter, theme) &&
    !missingLang(highlighter, lang)
  )
}

export async function loadHighlighter(
  lang: string,
  theme: EditorTheme
) {
  if (highlighterPromise === null) {
    setCDN("https://unpkg.com/shiki/")
    highlighterPromise = getHighlighter({
      theme: theme as IShikiTheme,
      langs: [lang as Lang], // TODO change lang from string to Lang
    })
  }

  if (highlighter === null) {
    highlighter = await highlighterPromise
  }
  if (missingTheme(highlighter, theme)) {
    await highlighter.loadTheme(theme as IShikiTheme)
  }
  if (missingLang(highlighter, lang)) {
    await highlighter.loadLanguage(lang as Lang)
  }
}

export function highlightOrPlaceholder(
  code: string,
  lang: string,
  theme: EditorTheme
) {
  if (isHighlighterReady(lang, theme)) {
    return getCodeLines(highlighter!, code, lang, theme)
  } else {
    return tokenizePlaceholder(code)
  }
}

function highlight(
  code: Tween<string>,
  lang: string,
  theme: EditorTheme
) {
  // assumes highlighter isReady

  const { fg, bg } = highlighter!.getTheme(theme.name)

  const lines = mapWithDefault(code, "", code =>
    getCodeLines(highlighter!, code, lang, theme)
  )

  return { lines, bg, fg }
}

function getCodeLines(
  highlighter: Highlighter,
  code: string,
  lang: string,
  theme: EditorTheme
): CodeLine[] {
  const lines = highlighter.codeToThemedTokens(
    code,
    lang,
    theme.name,
    {
      includeExplanation: false,
    }
  )

  return lines.map(line =>
    line.map(token => [
      token.content,
      { style: getStyle(token) },
    ])
  )
}

function missingTheme(
  highlighter: Highlighter,
  theme: EditorTheme
) {
  return !highlighter
    .getLoadedThemes()
    .some(t => t === theme.name)
}

function missingLang(
  highlighter: Highlighter,
  lang: string
) {
  return !highlighter
    .getLoadedLanguages()
    .some(l => l === lang)
}

const FONT_STYLE_TO_CSS = {
  [FontStyle.NotSet]: {},
  [FontStyle.None]: {},
  [FontStyle.Italic]: { fontStyle: "italic" },
  [FontStyle.Bold]: { fontWeight: "bold" },
  [FontStyle.Underline]: { textDecoration: "underline" },
}

function getStyle(token: IThemedToken) {
  const fontStyle = token.fontStyle
    ? FONT_STYLE_TO_CSS[token.fontStyle]
    : {}
  return {
    color: token.color,
    ...fontStyle,
  }
}

type ThemedCode = {
  lines: IThemedToken[][]
  fg: string
  bg: string
}

export function tokenizePlaceholder(
  code: string
): CodeLine[] {
  const lines = code.split("\n")
  return lines.map(line => [[line, {}]])
}

export function getBasicThemedCode(
  code: string,
  theme: EditorTheme
): ThemedCode {
  const lines = code.split("\n")
  const { fg, bg } = getThemeDefaultColors(theme)
  return {
    lines: lines.map(line => [{ content: line }]),
    fg,
    bg,
  }
}
