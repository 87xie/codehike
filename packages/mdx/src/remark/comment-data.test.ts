import { describe, it, expect, test } from "vitest"
import { highlight } from "../highlighter"
import { getCommentData } from "./comment-data"

test.each([
  "// focus",
  "       // focus",
  "foo // bar",
  "const x = 1",
  "  // box[6:9]",
  "  // bg(1:3) #222",
])("js | %s", async code => {
  expect(await getLineData("js", code)).toMatchSnapshot()
})

test.each([
  "# focus",
  "       # focus",
  "foo # bar",
  " x = '#333'",
  " # box[6:9]",
  "  # bg(1:3) #222",
])("python | %s", async code => {
  expect(
    await getLineData("python", code)
  ).toMatchSnapshot()
})

async function getLineData(lang: string, line: string) {
  return getCommentData(await asLine(lang, line), lang)
}

async function asLine(lang: string, line: string) {
  const code = await highlight({
    code: line,
    lang,
    theme: {},
  })
  return code.lines[0]
}
