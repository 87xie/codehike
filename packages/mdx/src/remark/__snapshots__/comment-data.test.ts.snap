// Vitest Snapshot v1

exports[`js |        // focus 1`] = `
{
  "data": "",
  "focusString": "",
  "key": "focus",
}
`;

exports[`js |   // bg(1:3) #222 1`] = `
{
  "data": "#222",
  "focusString": "(1:3)",
  "key": "bg",
}
`;

exports[`js |   // box[6:9] 1`] = `
{
  "data": "",
  "focusString": "[6:9]",
  "key": "box",
}
`;

exports[`js | // focus 1`] = `
{
  "data": "",
  "focusString": "",
  "key": "focus",
}
`;

exports[`js | const x = 1 1`] = `{}`;

exports[`js | foo // bar 1`] = `{}`;

exports[`python |        # focus 1`] = `
{
  "data": "",
  "focusString": "",
  "key": "focus",
}
`;

exports[`python |   # bg(1:3) #222 1`] = `
{
  "data": "#222",
  "focusString": "(1:3)",
  "key": "bg",
}
`;

exports[`python |  # box[6:9] 1`] = `
{
  "data": "",
  "focusString": "[6:9]",
  "key": "box",
}
`;

exports[`python |  x = '#333' 1`] = `{}`;

exports[`python | # focus 1`] = `
{
  "data": "",
  "focusString": "",
  "key": "focus",
}
`;

exports[`python | foo # bar 1`] = `{}`;
